File: api/src/interview_copilot/domain/questions.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class TopicData(BaseModel):
    id: UUID
    slug: str
    name: str


class SourceData(BaseModel):
    title: str
    url: str
    publisher: str


class QuestionEvidenceData(BaseModel):
    section_key: str
    heading_path: list[str]
    quote: str


class QuestionSummary(BaseModel):
    id: UUID
    slug: str
    title: str
    prompt: str
    difficulty: str
    question_type: str
    topics: list[TopicData]
    framework: str = "technical"
    source_document_id: UUID | None = None
    source_document_name: str | None = None
    source_document_version: int | None = None


class QuestionDetail(QuestionSummary):
    intent: str
    answer_outline: list[str]
    common_mistakes: list[str]
    sources: list[SourceData]
    content_markdown: str = ""
    editable: bool = False
    evidence: list[QuestionEvidenceData] = Field(default_factory=list)


class QuestionDocumentSummary(BaseModel):
    id: UUID
    filename: str
    media_type: str
    version: int
    status: str
    warnings: list[str]
    coverage_ratio: float = Field(ge=0, le=1)
    section_count: int = Field(ge=0)
    covered_section_count: int = Field(ge=0)
    question_count: int = Field(ge=0)
    knowledge_point_count: int = Field(default=0, ge=0)
    covered_knowledge_point_count: int = Field(default=0, ge=0)
    suggested_question_count: int = Field(default=0, ge=0, le=100)
    requested_question_limit: int = Field(default=30, ge=10, le=100)
    created_at: datetime
    updated_at: datetime


class QuestionImportResult(BaseModel):
    document: QuestionDocumentSummary
    questions: list[QuestionDetail]
    warnings: list[str] = Field(default_factory=list)


class QuestionSetSummary(BaseModel):
    id: UUID
    name: str
    kind: str
    status: str
    target_count: int = Field(ge=0, le=100)
    question_count: int = Field(ge=0)
    document_id: UUID | None = None
    document_name: str | None = None
    knowledge_point_count: int = Field(default=0, ge=0)
    covered_knowledge_point_count: int = Field(default=0, ge=0)
    created_at: datetime
    updated_at: datetime


class QuestionSetDetail(QuestionSetSummary):
    questions: list[QuestionSummary]


class CitationData(BaseModel):
    index: int
    title: str
    url: str | None = None
    quote: str


class QuestionChatAnswer(BaseModel):
    answer_markdown: str
    citations: list[CitationData]
    conversation_id: UUID


class QuestionChatMessageData(BaseModel):
    role: str = Field(pattern="^(user|assistant)$")
    content: str
    citations: list[CitationData] = Field(default_factory=list)
    created_at: datetime


class QuestionChatHistory(BaseModel):
    conversation_id: UUID
    messages: list[QuestionChatMessageData]


class UserQuestionState(BaseModel):
    status: str = "unseen"
    bookmarked: bool = False
    note: str = ""
    review_interval_days: int = Field(default=0, ge=0, le=365)
    review_streak: int = Field(default=0, ge=0, le=10_000)
    last_reviewed_at: datetime | None = None
    review_due_at: datetime | None = None

File: web/src/lib/questions.ts
import type { components } from "./api-schema";
import { z } from "zod";

export type QuestionSummary = components["schemas"]["QuestionSummary"];
export type QuestionDetail = components["schemas"]["QuestionDetail"];
export type UserQuestionState = components["schemas"]["UserQuestionState"];
export type QuestionDocumentSummary = components["schemas"]["QuestionDocumentSummary"];
export type QuestionChatAnswer = components["schemas"]["QuestionChatAnswer"];
export type QuestionChatHistory = components["schemas"]["QuestionChatHistory"];
export type QuestionChatMessageData = components["schemas"]["QuestionChatMessageData"];

export const questionSetSummarySchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  kind: z.string(),
  status: z.string(),
  target_count: z.number().int().nonnegative(),
  question_count: z.number().int().nonnegative(),
  document_id: z.string().uuid().nullable(),
  document_name: z.string().nullable(),
  knowledge_point_count: z.number().int().nonnegative(),
  covered_knowledge_point_count: z.number().int().nonnegative(),
  created_at: z.string(),
  updated_at: z.string(),
});
export const questionSetDetailSchema = questionSetSummarySchema.extend({
  questions: z.array(z.lazy(() => questionSummaryObjectSchema)),
});
export type QuestionSetSummary = z.infer<typeof questionSetSummarySchema>;
export type QuestionSetDetail = z.infer<typeof questionSetDetailSchema>;

const topicSchema = z.object({ id: z.string().uuid(), slug: z.string(), name: z.string() });
const citationSchema = z.object({ index: z.number().int().positive(), title: z.string(), url: z.string().nullable(), quote: z.string() });

const questionSummaryObjectSchema = z.object({
  id: z.string().uuid(),
  slug: z.string(),
  title: z.string(),
  prompt: z.string(),
  difficulty: z.string(),
  question_type: z.string(),
  topics: z.array(topicSchema),
  framework: z.string(),
  source_document_id: z.string().uuid().nullable(),
  source_document_name: z.string().nullable(),
  source_document_version: z.number().int().nullable(),
});
export const questionSummarySchema: z.ZodType<QuestionSummary> = questionSummaryObjectSchema;

export const questionDetailSchema: z.ZodType<QuestionDetail> = questionSummaryObjectSchema.extend({
  intent: z.string(),
  answer_outline: z.array(z.string()),
  common_mistakes: z.array(z.string()),
  sources: z.array(z.object({ title: z.string(), url: z.string(), publisher: z.string() })),
  content_markdown: z.string(),
  editable: z.boolean(),
  evidence: z.array(z.object({ section_key: z.string(), heading_path: z.array(z.string()), quote: z.string() })),
});

export const questionDocumentSchema: z.ZodType<QuestionDocumentSummary> = z.object({
  id: z.string().uuid(),
  filename: z.string(),
  media_type: z.string(),
  version: z.number().int().positive(),
  status: z.string(),
  warnings: z.array(z.string()),
  coverage_ratio: z.number().min(0).max(1),
  section_count: z.number().int().nonnegative(),
  covered_section_count: z.number().int().nonnegative(),
  question_count: z.number().int().nonnegative(),
  knowledge_point_count: z.number().int().nonnegative(),
  covered_knowledge_point_count: z.number().int().nonnegative(),
  suggested_question_count: z.number().int().min(0).max(100),
  requested_question_limit: z.number().int().min(10).max(100),
  created_at: z.string(),
  updated_at: z.string(),
});

export const questionChatAnswerSchema: z.ZodType<QuestionChatAnswer> = z.object({
  answer_markdown: z.string(),
  citations: z.array(citationSchema),
  conversation_id: z.string().uuid(),
});

export const questionChatHistorySchema: z.ZodType<QuestionChatHistory> = z.object({
  conversation_id: z.string().uuid(),
  messages: z.array(z.object({
    role: z.enum(["user", "assistant"]),
    content: z.string(),
    citations: z.array(citationSchema),
    created_at: z.string(),
  })),
});

export const userQuestionStateSchema: z.ZodType<UserQuestionState> = z.object({
  status: z.string(),
  bookmarked: z.boolean(),
  note: z.string(),
  review_interval_days: z.number().int().nonnegative(),
  review_streak: z.number().int().nonnegative(),
  last_reviewed_at: z.string().nullable(),
  review_due_at: z.string().nullable(),
});

export const QUESTION_INTERVIEW_SELECTION_KEY = "interview-copilot.question-selection.v1";
export const QUESTION_COACHING_SELECTION_KEY = "interview-copilot.coaching-question-selection.v1";
